### PR TITLE
Search & filter assets by tag README

### LINKED ISSUE #918: Create README: add label to asset; search & filter by label
There are several use cases where we want to retrieve data NFTs (or datatokens) by searching across all of them, and filtering down by label.

This should be easy for ocean.py users to figure out.

TODO for this issue: add a README where the user:
- creates a few data assets. Each asset has >= 1 labels
- searches & finds the assets by filtering by label

Potentially helpful resources:
- [df-py "Filter non-market assets" PR](https://github.com/oceanprotocol/df-py/pull/292) uses the patterns that we can use here.
- Base resource: https://docs.oceanprotocol.com/core-concepts/did-ddo 
- Another example: [https://1520763098-files.gitbook.io/~/files/v0/b/gitbook-x-prod.appspot.com/o/spaces%2FzQl[…]23f36dcef441e8fd76fdf6b6%2Fddo-flow%20(1).png?alt=media](https://1520763098-files.gitbook.io/~/files/v0/b/gitbook-x-prod.appspot.com/o/spaces%2FzQlpIJEeu8x5yl0OLuXn%2Fuploads%2Fgit-blob-fab9a027b0a25d4023f36dcef441e8fd76fdf6b6%2Fddo-flow%20(1).png?alt=media)

### PR DESCRIPTION
**Pull request by mariacarmina**
Fixes #918  .

Changes proposed in this PR:

- created README that implements the functionality;
- tested on ganache;
- updated `ocean.py` with the filtering function.

### CODE REVIEW

**Comment by mariacarmina** [comment]
Should I create tests in the suite for searching on remote networks as well?

**Comment by calina-c** [comment]
> Should I create tests in the suite for searching on remote networks as well?

No, I don't think that is necessary. This functionality is about creating assets and filtering them with Aquarius. We already test asset creation on remote networks, and asset filtering is not something network-dependent.

**Review by calina-c — APPROVED** [review]

**Review by calina-c — APPROVED** [review]

**Comment by calina-c** [comment]
@mariacarmina please wait for @trentmc to review as well. From my side, it looks good.

File: READMEs/search-and-filter-assets.md
Comment (trentmc) [inline READMEs/search-and-filter-assets.md:83]: Create() function already waits for Aquarius by default. 

Is that not enough?

If no, let's report another issue.

If yes, then remove the extra 5 s wait.

File: READMEs/search-and-filter-assets.md
Comment (trentmc) [inline READMEs/search-and-filter-assets.md:34]: We need to avoid having 50 lines of boilerplate. There's been great effort to fix this in ocean.py in the previous 8 weeks. I am loathe to reintroduce it.

And in general this is also a big DRY violation. We have to be working to reduce DRY violations, not the other way around.)

Solution: use "create_url_asset()" which is one line.

Then edit the ddo's tags via the README. Should be just 1-3 lines too.

That's it.

File: READMEs/search-and-filter-assets.md
Comment (trentmc) [inline READMEs/search-and-filter-assets.md:93]: There are many possible ways that people might search and filter assets. By tag is one. 

I think it will be better if we directly expose the 2-liner that you use to search and filter the assets. That will provide a better reference for them, to have other arbitrary searches.

(And IMO remove the new python method. It converts two lines to one, but obfuscates much greater flexibility.)

File: tests/integration/ganache/test_search_and_filter_assets.py
Comment (trentmc) [inline tests/integration/ganache/test_search_and_filter_assets.py:10]: Why have this function at all if we are testing the readme? 

We cab definitely remove it if we remove the new method.

In general it's also a big DRY violation. We have to be working to reduce DRY violations, not the other way around.

**Review by trentmc — CHANGES_REQUESTED** [review]

File: READMEs/search-and-filter-assets.md
Comment (mariacarmina) [reply READMEs/search-and-filter-assets.md:34]: I need to update the metadata which contains `tags` key, that's why I expanded the functionality, should I update `_create1` function in order to receive `metadata` as parameter?

File: READMEs/search-and-filter-assets.md
Comment (mariacarmina) [reply READMEs/search-and-filter-assets.md:83]: I will remove it.

File: tests/integration/ganache/test_search_and_filter_assets.py
Comment (mariacarmina) [reply tests/integration/ganache/test_search_and_filter_assets.py:10]: I got used when we created separate tests for README functionality, I'll make sure to delete it.

File: READMEs/search-and-filter-assets.md
Comment (trentmc) [reply READMEs/search-and-filter-assets.md:34]: > should I update _create1 function in order to receive metadata as parameter

No. Edit it after. That's why: "**edit** the ddo's tags". 

It's useful to show the user how to do this too.

File: READMEs/search-and-filter-assets.md
Comment (mariacarmina) [reply READMEs/search-and-filter-assets.md]: Should I mention that the `tag` is one option for filtering assets in a separate comment? BTW I have removed the function `search_asset_by_tag`.

File: READMEs/search-and-filter-assets.md
Comment (trentmc) [inline READMEs/search-and-filter-assets.md:54]: Tweak: make this one line

File: READMEs/search-and-filter-assets.md
Comment (trentmc) [inline READMEs/search-and-filter-assets.md:47]: Tweak: remove empty line

File: READMEs/search-and-filter-assets.md
Comment (trentmc) [inline READMEs/search-and-filter-assets.md:46]: Tweak: change
```python
for i in range(len(tags))
```

to 
```python
for tag in tags
```

And accordingly, below, change `tags[i]` to `tag`.

File: READMEs/search-and-filter-assets.md
Comment (trentmc) [inline READMEs/search-and-filter-assets.md:56]: Tweak: remove the part `New metadata is: {updated_asset.metadata}`. Why: it's unneeded, and would be a large printout which hinders usability.

File: READMEs/search-and-filter-assets.md
Comment (trentmc) [inline READMEs/search-and-filter-assets.md:55]: Tweak: for this line and below, rename `updated_asset` -> `asset`. Why: Creating a new variable makes things more error prone.

**Review by trentmc — APPROVED** [review]
In general, this is a great improvement! You responded to my other feedback appropriately. What's left is just a few tweaks. So I'm approving, assuming you make the tweaks before you merge. Thanks!

File: READMEs/search-and-filter-assets.md
Comment (trentmc) [reply READMEs/search-and-filter-assets.md]: It's fine as-is, since the code already makes it clear that they can do other sorts of filters.

File: READMEs/search-and-filter-assets.md
Comment (mariacarmina) [reply READMEs/search-and-filter-assets.md]: I thought that would create a confusion in readability because I am calling the old `asset` in the update function. I have changed as you purposed.

**Pull request merged**
